### ISSUE TITLE
Plugin Display Conflict

### ISSUE THREAD

**Issue by Cerothen** [ryanblenis/MeshCentral-ScriptTask#1]
Great plugin along with your other plugin RoutePlus.

When trying to load both of them at the same time the display for one overrides the other one depending on the order they were enabled.

EG. Loading the route plus plugin first will have its display override the Script task display. Loading the script task plugin first will have the script display override the route plus display.

Is there a way to prevent this from happening? I understand plugins are a pretty new thing in meshcentral2

thanks.

**Comment by ryanblenis** [comment]
Hey @Cerothen,

I can't seem to replicate this. Can you let me know what browser / OS you're using as well as step by step instructions on how to replicate? Does this pertain to installing the plugins, or pulling up the device's "Plugins" tab as well as the "My Account" -> RoutePlus tab? Thank you.

**Comment by Cerothen** [comment]
Chrome: Version 79.0.3945.88 (latest)
MeshCentral: Version 0.4.6-v (latest npm)

So it seems that whatever plugin is accessed first following a server restart is the one that shows on both views. Below is an example with only one plugin enabled (initial) it works as expected, when enabling the second plugin and accessing the "routeplus" plugin area first results in the device level view that would normally be the "ScriptTask" view to be the "routeplus" view though not functional.

Let me know if I haven't provided enough detail.

Initial:
![image](https://user-images.githubusercontent.com/9008757/71707289-1acb0800-2db7-11ea-9358-cf71a2aaecd7.png)
![image](https://user-images.githubusercontent.com/9008757/71707340-3fbf7b00-2db7-11ea-928a-792f182261c2.png)
* No "Plugin" section under a device

Enabling ScriptTasks:
![image](https://user-images.githubusercontent.com/9008757/71707407-a8a6f300-2db7-11ea-925f-cdc6a9ae70c4.png)
![image](https://user-images.githubusercontent.com/9008757/71707433-ce33fc80-2db7-11ea-8ac7-91468bdb11bd.png)
![image](https://user-images.githubusercontent.com/9008757/71707424-c1afa400-2db7-11ea-92e1-39c24e3deb80.png)

**Comment by ryanblenis** [comment]
Hmm, I still can't seem to replicate it. Here's what I'm doing:

1. Disable ScriptTask, leave RoutePlus enabled.
2. Restart the server
3. Go to My Account -> RoutePlus
4. Go to My Server -> Plugins -> enable ScriptTask
5. Refresh the page (probably something that needs an update in MeshCentral's plugin system, as the tab doesn't show up in devices yet if I don't refresh here)
6. Go to My Account -> RoutePlus
7. Go to a device -> Plugins -> ScriptTask (I see ScriptTask, but I'm assuming you see RoutePlus here)

Let me know if I'm doing something other than you here. FYI, please also try with your browser console open and see if there are any errors. The moment you navigate away from RoutePlus, the iframe should be destroyed, so I can't imagine how it would be showing by the time you get to a device. Thank you.

Also, if you click on the ScriptTask header in the Device -> Plugins section, does that change anything? Thank you.

**Comment by Cerothen** [comment]
So I did a little bit more investigating and determined that there was a missing module.

Running ```npm install --save semver``` from the project directory resolved the issue and seems to allow me to access both plugins simultaneously. The web console does still show ```Error loading plugin handler (TypeError: Cannot read property '_id' of undefined) - (index):1672``` which is this line ```try { pluginHandler[message.plugin][message.method](server, message); } catch (e) { console.log('Error loading plugin handler ('+ e + ')'); }```

**Comment by ryanblenis** [comment]
Ah, I see! Interesting. It appears semver is required by devDependencies (of other dependencies) in MeshCentral but not explicitly required. Hence if your NODE_ENV is not development, semver will not be present. I can see how this causes the plugins page to say "Checking..." in your screenshots versus actual version numbers, as the MeshCentral pluginHandler uses semver to compare. Not sure exactly how it affects iframes overlapping, but I'm guessing because the server chokes and never finishes certain calls after that. Thank you for digging in and finding that! I'll work with MeshCentral to make sure that is updated so this is not an issue moving forward.

**Comment by ryanblenis** [comment]
Semver has now been added as a module if plugins are enabled in the MeshCentral project. 
As for the web console error, the line number can vary depending on installed plugins. My best guess is a "My Computer" is not selected for the RoutePlus addon, in which case that is fine. However if you change 
`console.log('Error loading plugin handler ('+ e + ')');`
to
`console.log('Error loading plugin handler ('+ e.stack + ')');` it will provide more information about which line is throwing the error. Thank you.

**Comment by ryanblenis** [comment]
Hi @Cerothen,

I found the issue with the additional console logs showing up and published a fix in a couple plugins. Closing this out now. Thank you.

**Comment by ryanblenis** [comment]
Just noting here that I came across this original issue today (plugin page display conflict). The issue is that MeshCentral is currently set to change the 'views' directory dynamically based on the plugin being called. With multiple plugins this is providing a race condition where another plugin can reset the 'views' directory before the content is loaded by the first plugin. To overcome this, I've added absolute paths to all my plugins.
